### PR TITLE
Spike auth

### PR DESCRIPTION
**Pull request by roblinksdata**

### CODE REVIEW

File: app/server.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    return response
+
+
+@app.route('/')
+async def home(request: Request):
+    """
+    Super simple endpoint, returns some json
+    """
+    auth = Auth(ac, request, logger)
+    return json({"you're logged in as an admin": auth.has_admin()}, status=200)
```
Comment (robons) [inline app/server.py:37]: Yeah, as I think you've been mentioning it'd be good to be able to use a decorator like `@authorize(roles=["admin"])` to ensure only permitted individuals can access protected end-points.

File: auth-spike.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+### Rules
+
+Auth0 [rules](https://auth0.com/docs/customize/rules) allow you to extend the information returned from the `/userinfo` endpoint, so **after** authentication, but **before** said userinfo is returned to the app.
+
+In this case I've set a `rule` on an [https://auth0.com/](https://auth0.com/) account to add an `admin` role under the roles namespace (see configuration.ini) but only for authenticated google `@gsscogs.uk` users.
+
+_Note - the roles namespace defined in the apps cofiguration **must** match the same one used in the auth `rule`._ 
+
+In production the `rule`'s would likely be more nuanced than this (some variation of calling off to a database to get a given users roles) but the approach would/could be the same or similar, i.e
```
Comment (robons) [inline auth-spike.md:70]: As discussed, it'd be good to try and get the [Auth0 RBAC](https://auth0.com/docs/authorization/rbac) stuff up and running properly.

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[[source]]
```
Comment (robons) [inline Pipfile:1]: Can we use poetry instead of pipenv please? We've encountered so many problems with pipenv taking an enternity on csvcubed.

**Review by roblinksdata — COMMENTED** [review]
Thanks for getting this spike done. It'd be great to take this code, start making some improvements and additions and develop it into something we can deploy.

File: app/auth.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            scope='openid',
+            redirect_uri=self.cfg.redirect_uri
+        )
+
+        self.cookie = request.cookies.get('user', None)
+        if not self.cookie:
+            self.pristine()
+
+
+    def encrypt_cookie(self, sub_dict):
```
[the remark this answers is not in the document]
Comment (thomasmastersmethods) [reply app/auth.py:34]: agree would be good practise - also, perhaps add a 'refresh token' function, would be benficial when hooking this up with the API

File: app/utils/configuration/auth.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        self.roles_namespace = cfg.get('AUTH', 'roles_namespace')
+
+        # Things that started as env vars
+        self.encryption_key = cfg.get('ENV', 'encryption_key')
+        self.client_id = cfg.get('ENV', 'oauth_client_id')
+        self.client_secret = cfg.get('ENV', 'oauth_client_secret')
+        self.redirect_uri = cfg.get('ENV', 'oauth_redirect_uri')
+
+        # Extrapolated
+        self.access_token_url = f'https://{oauth_domain}/oauth/token'
```
[the remark this answers is not in the document]
Comment (thomasmastersmethods) [reply app/utils/configuration/auth.py:26]: RE: client_secret, client_id, encryption_key - rather than having this in 'ENV' when we go to prod, perhaps have this as a keystorage solution in the cloud when we get further along with it

File: app/auth.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    def has_admin(self) -> (bool):
+        """
+        Does the current user cookie contain an access token that
+        identifies them as having then "admin" role under the
+        roles_namespace key.
+        """
+        return self._has_role("admin")
+
+
+    def pristine(self):
```
Comment (thomasmastersmethods) [inline app/auth.py:234]: pep8 - too many blank lines between functions

File: app/server.py (hunk 1 above)
Comment (robons) [reply app/server.py:37]: @thomasmastersmethods if you find yourself with some spare time this afternoon, do you want to try giving this a go?

File: app/auth.py (hunk 6 above)
Comment (robons) [reply app/auth.py:234]: indeed, let's use [black](https://pypi.org/project/black/) on this project though.

File: app/utils/configuration/auth.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomasmastersmethods) [reply app/utils/configuration/auth.py:26]: re: file naming convention - possibly think about changing, doesnt conflict with app/auth.py

File: app/server.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    Do the right things when writing a user cookie
+    along with a response.
+    """
+    response.cookies["user"] = cookie
+    response.cookies['user']['httponly'] = True
+    response.cookies['user']['secure'] = True
+    
+    # Note - split as it needs to be "localhost" not "localhost:<port>"
+    # TODO - use a library to make url, not nasty splits
+    response.cookies['user']['domain'] = f'{request.host.split(":")[0]}'
```
Comment (thomasmastersmethods) [inline app/server.py:26]: perhaps look at 'urllib' library

**Comment by mikeAdamss** [comment]
cheers for the feedback both, good stuff 👍 , will incorporate into follow up tasks

File: app/server.py (hunk 1 above)
Comment (thomasmastersmethods) [reply app/server.py:37]: Yes, if I get a window certainly take a look at it

File: app/utils/configuration/decorators.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+from functools import wraps
+
+
+def authorised(func):
+    def authoriser(name=None):
+        @wraps(func)
+        async def decorated_function(request, *args, **kwargs):
+            # if 'upload' in request.url: TO DO: mapping to permission/role
+            # system
+            response = f(request, *args, **kwargs)
```
Comment (roblinksdata) [inline app/utils/configuration/decorators.py]: @thomasmastersmethods: What is `f` in this context? Is it bound to anything? And where does this throw unauthenticated users out?

File: app/utils/configuration/main.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    section in the configuration.ini for default values, so we check for an env
+    var override for each field - using the ini default in its absense.
+
+    There's a fine line between convenience and obfiscation here, if we do it we may
+    need logging out of sanitised variables, warnings per default use, something
+    like that.... think.
+    """
+
+    # Police current MVP approach
+    assert ["ENV"] not in config.sections(), (
```
Comment (roblinksdata) [inline app/utils/configuration/main.py]: So if they do have "ENV", you'll complain that they must have the thing they have? I'm confused.

Comment (mikeAdamss) [reply app/utils/configuration/main.py]: this one is my bad from quick and dirty spike code 😄 , it was _aimed_ at enforce there is always an _empty_ section called `ENV` (I wanted to basically reserve it for dynamic population from env vars only - save future people tripping up on that bit of fuctionalty.).  

Still waking up but I think it needs to be `"ENV"` not `not ["ENV"]` to make sense.

File: tests/test_authentication_decorator.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
+from sanic import Sanic
+from websockets import WebSocketClientProtocol
+import pytest
+import pytest_asyncio
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: unused import.

File: app/utils/configuration/handlers.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+from sanic.response import json
+from exceptions import (
+    AccessDeniedError,
+    ConfigurationConflictError,
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: This isn't used, your IDE should be _screaming_ that at you, what IDE are you using?

This is also an import error so this codebase cannot be ran without exceptions, should it be in pr?

File: app/utils/configuration/handlers.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        lambda r, e: json({Handler.default_message_key: str(e)}, 403)
+    )
+
+    @classmethod
+    def _set_error_handlers(cls, app):
+        app.error_handler.add(
+            NoAuthorizationError,
+            cls.handler.no_authorization)
+        app.error_handler.add(
+            ExpiredSignatureError,
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: `ExpiredSignatureError`, this doesn't exist, again your IDE should be having fits about it, also - this code has never been ran (or even imported) otherwise you'd have been buried in exceptions.

File: app/utils/configuration/handlers.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+            cls.handler.expired_signature)
+        app.error_handler.add(InvalidHeaderError, cls.handler.invalid_header)
+        app.error_handler.add(InvalidTokenError, cls.handler.invalid_token)
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: `InvalidTokenError` doesn't exist either, see above comment.

File: app/utils/configuration/handlers.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
+from sanic.response import json
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: _general point on this file not this line._

I can sort of see why we wanted to write an error handler (so our tests can check for the expected errors) but (a) we're not doing that so I'm not convinced it has a place in this pr and (b) is `/configuration` the correct place?

File: app/utils/configuration/exceptions.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
+class JWTExtendedException(Exception):
```
Comment (mikeAdamss) [inline app/utils/configuration/exceptions.py]: _general point on this file not this line._

I see where you're coming from but I don't think the `/configuration` module is the right place for this. errors are nether configurable nor configuration. I'd be tempted by something like `app/exceptions/oidc.py` (or some variation of, I'm pretty bad with naming conventions and structuring) so we can keep errors centralised but also grouped for organisational purposes.

I'd also be pretty surprised if some of these didn't already exist, the `jwt` library seems likely to already have a `JWTDecodeError` for example.

File: app/auth.py [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    also holds associated authentication and authorization methods.
+    """
+
+    def __init__(self, cfg: AuthConfig, request: Request, logger):
+        self.cfg = cfg
+        self.logger = logger
+        self.request = request
+   
+        self.client = OAuth2Session(
+            self.cfg.get('client_id', None),
```
Comment (mikeAdamss) [inline app/auth.py]: Not sure why we changed this to dictionary style but `AuthConfig` is a class not a dictionary (see type hint) and all the `get` and default logic is handled during its own instantiation.

File: app/utils/configuration/decorators.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+from handlers import Handler
+
+
+def authorised(func):
+    def authoriser(name=None):
+        @wraps(func)
+        async def decorated_function(request, *args, **kwargs):
+            # TO DO: mapping to permission/role from user id
+            response = f(request, *args, **kwargs)
+            if isawaitable(response):
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: At a guess, I think you might be missing `from inspect import isawaitable` in your imports. Again, (a) your IDE should be having a fit and (b) this code has never been ran - should it be in pr as finished?

File: app/utils/configuration/decorators.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+from sanic.response import redirect
+from handlers import Handler
+
+
+def authorised(func):
+    def authoriser(name=None):
+        @wraps(func)
+        async def decorated_function(request, *args, **kwargs):
+            # TO DO: mapping to permission/role from user id
+            response = f(request, *args, **kwargs)
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: rob has already asked about this in an unaddressed pr comment from last week - try and address review comments before putting things into pr as finished please.

In this case `f` (I believe) is `func` but misnamed, is that right? did you copy paste this code from somewhere?

File: app/utils/configuration/decorators.py [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+            if isawaitable(response):
+                response = await response
+                if 'name is not authorised':
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: The boolean value of a string in python is always `True` so the else part of this conditional is unreachable.

File: app/utils/configuration/decorators.py [hunk 20]
```diff
@@ -0,0 +1,24 @@
+from functools import wraps
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: _general point on this file not this line._

Not convinced this should be in the configuration module.

File: app/utils/configuration/decorators.py [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+from sanic.response import redirect
+from handlers import Handler
+
+
+def authorised(func):
+    def authoriser(name=None):
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: I don't understand this, is `name` an alias of `role`?

More broadly I don't understand how any of this is connected to wrapping the `Auth` class with a decorator, shouldn't the `Auth` class be involved somewhere?

File: tests/test_authentication_decorator.py [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+import requests
+import json
+from logging import Logger
+from mock import Mock
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: Do you mean `from unittest.mock import Mock` ? It's been included in base python since 3.3.

File: tests/test_authentication_decorator.py [hunk 23]
```diff
@@ -0,0 +1,184 @@
+from sanic import Sanic
+from websockets import WebSocketClientProtocol
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: unused import

File: tests/test_authentication_decorator.py [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+import pytest
+import pytest_asyncio
+import requests
+import json
+from logging import Logger
+from mock import Mock
+from app.auth import Auth
+from app.utils.configuration.decorators import authorised
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: You'll probably need to add `__init__.py` to `/app` to make these work.

File: app/utils/configuration/decorators.py [hunk 25]
```diff
@@ -0,0 +1,24 @@
+from functools import wraps
+from sanic.response import redirect
+from handlers import Handler
```
Comment (mikeAdamss) [inline app/utils/configuration/decorators.py]: broken import, might need to be `from .handlers import Handler`

File: app/utils/configuration/handlers.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+from exceptions import (
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: broken import, might need to be `from .exceptions import (`

File: tests/test_authentication_decorator.py [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+    # @pytest.fixture
+    @pytest.mark.asyncio
+    def test_no_user(self, app):
+        """
+        Test user not authenticated
+        """
+
+        func = Mock()
+        decorated_auth_func = authorised(func)
+        request, resp = app.test_client.get('/')
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: `app.test_client` uses `sanic_testing` which isn't installed, you'll need to add it to `tool.poetry.dev-dependencies` in `pyproject.toml` then run make (you might need to remove the `.lock` before you `make` as well, unsure.

you'll need to merge in `develop` to get poetry in as per comments of last week.

File: tests/test_authentication_decorator.py [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+from logging import Logger
+from mock import Mock
+from app.auth import Auth
+from app.utils.configuration.decorators import authorised
+
+
+
+class TestDecorator:
+
+    @pytest.mark.asyncio
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: both of these decorators seem result in depreciation warnings, that probably needs to be addressed.

File: tests/test_authentication_decorator.py (hunk 27 above)
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: how is this line connected to the line above or below it?

File: tests/test_authentication_decorator.py [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+    def test_no_user(self, app):
+        """
+        Test user not authenticated
+        """
+
+        func = Mock()
+        decorated_auth_func = authorised(func)
+        request, resp = app.test_client.get('/')
+        dummy_response = decorated_auth_func(name='dummy_name')
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: why are we assigning to a variable we don't use?

File: tests/test_authentication_decorator.py [hunk 30]
```diff
@@ -0,0 +1,184 @@
...
+        assert not func.called
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: this is all really hard to follow but isn't `dummy_response` here a function not a response? Are we missing some variation of `dummy_response(request)` ? Otherwise are we just confirming that a function we never tried to call isn't called?

File: tests/test_authentication_decorator.py [hunk 31]
```diff
@@ -0,0 +1,184 @@
...
+
+    @pytest.mark.asyncio
+    def test_user_authenticated(self, app):
+        """
+        Test user authenticated
+        """
+        func = Mock(return_value='ok')
+        decorated_auth_func = authorised(func)
+        dummy_request = app.test_client.request
+        dummy_response = decorated_auth_func(dummy_request)
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: in the previous test we passed in a name variable, in this one we're passing in a http request ... I don't understand.

File: tests/test_authentication_decorator.py [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+        assert dummy_response._mock_return_value == 'ok'
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: As per previous test your `dummy_response` is a function not a response, the `decorated_function` part of your decorator is _not_ being called by your test logic. I suspect this is a test that can't fail.

File: tests/test_authentication_decorator.py [hunk 33]
```diff
@@ -0,0 +1,184 @@
...
+        assert resp_json == {"GET": True}
+
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_post(self, app):
+        """
+        Test POST request
+        """
+        resp = await app.test_client.post('/test_post')
+        assert resp.status_code == 200
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: `resp` is a tuple and doesn't have attributes.

File: tests/test_authentication_decorator.py [hunk 34]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        resp = await app.test_client.get('/test_get')
+        assert resp.status_code == 200
+        resp_json = await resp.json()
+        import pdb; pdb.set_trace()
+        assert resp_json == {"GET": True}
+
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_post(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see all comments for `test_fixture_test_client_get`

File: tests/test_authentication_decorator.py [hunk 35]
```diff
@@ -0,0 +1,184 @@
...
+        Test user authenticated
+        """
+        func = Mock(return_value='ok')
+        decorated_auth_func = authorised(func)
+        dummy_request = app.test_client.request
+        dummy_response = decorated_auth_func(dummy_request)
+        assert dummy_response._mock_return_value == 'ok'
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_get(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: you can't run this as `async`

File: tests/test_authentication_decorator.py [hunk 36]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        Test GET request
+        """
+        resp = await app.test_client.get('/test_get')
+        assert resp.status_code == 200
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see previous comment to fix this so it has attributes, once you do ... it won't be 200.

File: tests/test_authentication_decorator.py [hunk 37]
```diff
@@ -0,0 +1,184 @@
...
+        dummy_request = app.test_client.request
+        dummy_response = decorated_auth_func(dummy_request)
+        assert dummy_response._mock_return_value == 'ok'
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_get(self, app):
+        """
+        Test GET request
+        """
+        resp = await app.test_client.get('/test_get')
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: I believe this should be `request, resp = await app.test_client.get('/test_get')` otherwise `resp` is a tuple.

File: tests/test_authentication_decorator.py [hunk 38]
```diff
@@ -0,0 +1,184 @@
...
+        assert resp.status_code == 200
+        resp_json = await resp.json()
+        import pdb; pdb.set_trace()
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: what is this doing? also don't force two statements onto a single line in this way please

File: tests/test_authentication_decorator.py [hunk 39]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        Test POST request
+        """
+        resp = await app.test_client.post('/test_post')
+        assert resp.status_code == 200
+        resp_json = await resp.json()
+        assert resp_json == {"POST": True}
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_put(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see all comments for `test_fixture_test_client_get`

File: tests/test_authentication_decorator.py [hunk 40]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        Test PUT request
+        """
+        resp = app.test_client.put('/test_put')
+        assert resp.status_code == 200
+        resp_json = resp.json()
+        assert resp_json == {"PUT": True}
+
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_delete(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see all comments for `test_fixture_test_client_get`

File: tests/test_authentication_decorator.py [hunk 41]
```diff
@@ -0,0 +1,184 @@
...
+    @pytest.mark.asyncio
+    async def test_fixture_test_client_get(self, app):
+        """
+        Test GET request
+        """
+        resp = await app.test_client.get('/test_get')
+        assert resp.status_code == 200
+        resp_json = await resp.json()
+        import pdb; pdb.set_trace()
+        assert resp_json == {"GET": True}
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: I don't understand where this would be coming from? where is the endpoint `/test_get` defined? Where does this json come from?

File: tests/test_authentication_decorator.py [hunk 42]
```diff
@@ -0,0 +1,184 @@
...
+            auth_manager.cfg['client_id'] = "client_id"
+
+            @app.route("/protected", methods=["GET"])
+            async def protected(*args, **kwargs):
+                return json({}, 204)
+
+            yield app
+
+    @pytest.mark.asyncio
+    async def test_jwt_wrong_token(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: had to remove `async ` and `await` to get this to run, once I did it started failing with illegal header name.

File: tests/test_authentication_decorator.py [hunk 43]
```diff
@@ -0,0 +1,184 @@
...
+        """
+            Test 'incorrect' JWT token and 'forbidden' response
+        """
+        dummy_token = ''
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: this is why the illegal header name - you can't leave these blank, you're creating malformed http headers.

File: tests/test_authentication_decorator.py [hunk 44]
```diff
@@ -0,0 +1,184 @@
...
+        dummy_JWT_header_key = ''
+        dummy_JWT_header_prefix = ''
+
+        resp = await app.test_client.get(
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: `request, resp = `

File: tests/test_authentication_decorator.py [hunk 45]
```diff
@@ -0,0 +1,184 @@
...
+            '/protected',
+            headers={
+                dummy_JWT_header_key: f"{dummy_JWT_header_prefix} {dummy_token}"},
+        )
+
+        assert resp.status == 422
+
+    @pytest.mark.asyncio
+    def test_jwt_correct_token(self, app):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see comments for above test

File: tests/test_authentication_decorator.py [hunk 46]
```diff
@@ -0,0 +1,184 @@
...
+from app.utils.configuration.decorators import authorised
+
+
+
+class TestDecorator:
+
+    @pytest.mark.asyncio
+    @pytest.yield_fixture
+    def app(self):
+        import random
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: we don't want to `import random` every time we call this fixture.

File: tests/test_authentication_decorator.py [hunk 47]
```diff
@@ -0,0 +1,184 @@
...
+        app = Sanic(name=str('_' + str(random.randint(0, 10000))))
+
+        @app.route("/protected", methods=["GET"])
+        async def protected(*args, **kwargs):
+            return response.json({}, 204)
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: `response` hasn't been defined.

File: tests/test_authentication_decorator.py [hunk 48]
```diff
@@ -0,0 +1,184 @@
...
+        Test DELETE request
+        """
+        resp = app.test_client.delete('/test_delete')
+        assert resp.status_code == 200
+        resp_json = resp.json()
+        assert resp_json == {"DELETE": True}
+
+    @pytest.fixture
+    @pytest.mark.asyncio
+    def test_create_access_token(self, mocker):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: is this a fixture? What is mocker and where does it come from?

File: tests/test_authentication_decorator.py [hunk 49]
```diff
@@ -0,0 +1,184 @@
...
+        with Auth(dummy_cfg, requests, Logger) as auth_manager:
+            auth_manager.cfg['client_secret'] = "secret"
+            auth_manager.cfg['client_id'] = "client_id"
+            raw_token = auth_manager.get_access_token()
+
+            assert raw_token == "access_token"
+
+    @pytest.fixture
+    @pytest.mark.asyncio
+    def test_app_initialize(self, mocker):
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: is this fixture used by anything?

File: tests/test_authentication_decorator.py [hunk 50]
```diff
@@ -0,0 +1,184 @@
...
+                     'algorithm': ''}
+
+        mock_cfg = mocker.patch.object(Auth.self.client, '')
+        mock_cfg.return_value = 'dummy_client'
+        with Auth(dummy_cfg, requests, Logger) as auth_manager:
+            auth_manager.cfg['client_secret'] = "secret"
+            auth_manager.cfg['client_id'] = "client_id"
+            raw_token = auth_manager.get_access_token()
+
+            assert raw_token == "access_token"
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: I don't really follow this, so bear with:

We're patching `Auth.self.client` (isn't `Auth.self` redundant?) to replace `.client` with something equal to the str `dummy_client`

how would this result in `get_access_token` returning the text "access_token" ?

File: tests/test_authentication_decorator.py [hunk 51]
```diff
@@ -0,0 +1,184 @@
...
+
+        mock_cookies = mocker.patch.object(Auth.self.cookie, '')
+        mock_cookies.return_value = 'dummy_cookie'
+        with Auth(dummy_cfg, requests, Logger) as auth_manager:
+            auth_manager.cfg['client_secret'] = "secret"
+            auth_manager.cfg['client_id'] = "client_id"
+
+            @app.route("/protected", methods=["GET"])
+            async def protected(*args, **kwargs):
+                return json({}, 204)
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: this is json from `import json`, I suspect you'll want `from sanic import json`, very different things.

File: app/utils/configuration/handlers.py [hunk 52]
```diff
@@ -0,0 +1,61 @@
...
+    FreshTokenRequiredError,
+    InvalidHeaderError,
+    JWTDecodeError,
+    NoAuthorizationError,
+    RevokedTokenError,
+    WrongTokenError,
+)
+
+
+class Handler:
```
Comment (mikeAdamss) [inline app/utils/configuration/handlers.py]: `ExceptionHandler`possibly?

File: tests/test_authentication_decorator.py [hunk 53]
```diff
@@ -0,0 +1,184 @@
...
+        )
+
+        assert resp.status == 422
+
+    @pytest.mark.asyncio
+    def test_jwt_correct_token(self, app):
+        """
+            Test 'Correct' JWT token with 'success' response
+        """
+        dummy_token = ''
```
Comment (mikeAdamss) [inline tests/test_authentication_decorator.py]: see earlier points about malformed http headers.

**Review by mikeAdamss — COMMENTED** [review]
I think this one needs quite a lot of love, in terms of the auth decorator I mean (there are plenty of other bits from my original spike that need attention too, but that's a different task so I've kept to your changes only).

To begin with you probably need a better IDE setup, trying to merge with this amount of missing and/or incorrect imports shouldn't really be a thing.

As it stands both the main codebase and the tests are broken, I don't mean "failing" (in case of the tests) I mean broken, as in neither the new code nor the tests can be utilised as intended without exceptions being raised. I also don't really understand what we've got here - I thought we were wrapping the `Auth` class in a decorator, where is that happening?

I've tried to be thorough with this but tbh, a pretty significant amount of this feedback you could have identified yourself by running your code and/or tests before submitting them for review, I'd regard that as pretty reasonably standard practice to aim for tbh.

**Comment by thomasmastersmethods** [comment]
> I think this one needs quite a lot of love, in terms of the auth decorator I mean (there are plenty of other bits from my original spike that need attention too, but that's a different task so I've kept to your changes only).
> 
> To begin with you probably need a better IDE setup, trying to merge with this amount of missing and/or incorrect imports shouldn't really be a thing.
> 
> As it stands both the main codebase and the tests are broken, I don't mean "failing" (in case of the tests) I mean broken, as in neither the new code nor the tests can be utilised as intended without exceptions being raised. I also don't really understand what we've got here - I thought we were wrapping the `Auth` class in a decorator, where is that happening?
> 
> I've tried to be thorough with this but tbh, a pretty significant amount of this feedback you could have identified yourself by running your code and/or tests before submitting them for review, I'd regard that as pretty reasonably standard practice to aim for tbh.

Thanks Mike for the feedback - I was expecting some comments on this given the loose requirements, given the permissions/roles/user framework isnt yet established - though let me address the items raised

File: tests/test_authentication_decorator.py (hunk 10 above)
Comment (thomasmastersmethods) [reply tests/test_authentication_decorator.py]: fixed

File: app/utils/configuration/handlers.py (hunk 11 above)
Comment (thomasmastersmethods) [reply app/utils/configuration/handlers.py]: fixed, yes added this in as extras really

File: app/utils/configuration/handlers.py (hunk 12 above)
Comment (thomasmastersmethods) [reply app/utils/configuration/handlers.py]: fixed

File: app/utils/configuration/handlers.py (hunk 13 above)
Comment (thomasmastersmethods) [reply app/utils/configuration/handlers.py]: fixed

File: tests/test_authentication_decorator.py (hunk 50 above)
Comment (thomasmastersmethods) [reply tests/test_authentication_decorator.py]: yes had teething issues testing with async etc - let me look at this

**Comment by mikeAdamss** [comment]
I think there are some misunderstandings here tbh, if you have a look at the task: https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gss-cogs/linked-data-repository/10 we just need to move the existing Auth() classes logic into a decorator.

the code for checking whether someone already has "admin" role is already written in that Auth class, we need to move that class (in particular its `._has_role()` logic) into a decorator.

I think permissions frameworks are a bit of a red herring tbh, since the Auth class from the spike can check if someone has as "admin" role already (if you set up auth0 as per the google doc) that self same logic will work for any role right? there's no need to have a permissions framework to do that.

**Comment by mikeAdamss** [comment]
@thomasmastersmethods  - I've gone through and gotten things to working with an example decorator (see `app/utils/auth/decorator.py`).

do **finish what you're currently working on**, but when you come back to this ....

I'm not saying the example I've written is _the_ implementation (few hours work only) however:

* if you setup your `auth0` as per the word documentation I sent you previously....
* the endpoints commented as `EXAMPLE` in server.py have a working auth deocrator and follow all patterns required by: https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gss-cogs/linked-data-repository/10

from there the decorator task will be:

* fill in the blank decorator tests I've included, the EXAMPLE endpoints should give you some ideas.
* do any tidy up of my quick hackery you think is required (feel free to rewrite the decorator if you like - though I'd probably write the tests before you do).
* add any other tests you think are relevant for the decorator.
* anything you can add to `test_auth_manager.py` would be a bonus but don't lose yourself down another rabbit hole.
* remove the EXAMPLE tagged endpoints.

File: app/utils/configuration/handlers.py (hunk 11 above)
Comment (mikeAdamss) [reply app/utils/configuration/handlers.py]: not sure what's gone on here, but this wasn't fixed until I removed the code.

File: app/utils/configuration/handlers.py (hunk 12 above)
Comment (mikeAdamss) [reply app/utils/configuration/handlers.py]: same as other fixed one, is it possible you didn't commit these changes?

File: app/utils/configuration/handlers.py (hunk 13 above)
Comment (mikeAdamss) [reply app/utils/configuration/handlers.py]: wasn't fixed
